File: lesson_04/devynbenson/test_prime_function.js
function primeFunction(n) {
    if (!Number.isInteger(n) || n <= 1) 
        return false;
   
    if (n <= 3) 
        return n >= 2;              
    
    if (n % 2 === 0 || n % 3 === 0) 
        return false;

    // Check for divisors from 5 to √n
    for (let i = 5; i * i <= n; i += 6) {
        if (n % i === 0 || n % (i + 2) === 0) {
            return false;
        }
    }
    
    return true;
}


// Simple test framework
function test(description, testFunction) {
    try {
        testFunction();
        console.log(`PASS: ${description}`);
    } catch (error) {
        console.log(`FAIL: ${description}`);
        console.log(`ERROR: ${error.message}`);
    }
}

function assertEqual(actual, expected, message = "") {
    if (actual !== expected) {
        throw new Error(`Expected ${expected}, but got ${actual}. ${message}`);
    }
}

// Test Suite
console.log("Running JavaScript Prime Function Tests...\n");

// Test edge cases
test("Edge case: 1 (only one positive divisor)", () => {
    assertEqual(primeFunction(1), false);
});

test("Edge case: 0", () => {
    assertEqual(primeFunction(0), false);
});

test("Edge case: negative numbers", () => {
    assertEqual(primeFunction(-1), false);
    assertEqual(primeFunction(-5), false);
});

test("Edge case: non-integer input", () => {
    assertEqual(primeFunction(2.5), false);
    assertEqual(primeFunction(3.14), false);
});

// Test small primes
test("Small prime: 2 (only even prime)", () => {
    assertEqual(primeFunction(2), true);
});

test("Small primes: 3, 5, 7, 11, 13", () => {
    assertEqual(primeFunction(3), true);
    assertEqual(primeFunction(5), true);
    assertEqual(primeFunction(7), true);
    assertEqual(primeFunction(11), true);
    assertEqual(primeFunction(13), true);
});

// Test composite numbers
test("Composite: 4 (4 = 2×2)", () => {
    assertEqual(primeFunction(4), false);
});

test("Composite: 6 (6 = 2×3)", () => {
    assertEqual(primeFunction(6), false);
});

test("Composite: 9 (9 = 3×3)", () => {
    assertEqual(primeFunction(9), false);
});

test("Composite numbers: 8, 10, 12, 15", () => {
    assertEqual(primeFunction(8), false);
    assertEqual(primeFunction(10), false);
    assertEqual(primeFunction(12), false);
    assertEqual(primeFunction(15), false);
});

// Test larger primes
test("Larger primes: 17, 19, 23, 29, 31, 97", () => {
    assertEqual(primeFunction(17), true);
    assertEqual(primeFunction(19), true);
    assertEqual(primeFunction(23), true);
    assertEqual(primeFunction(29), true);
    assertEqual(primeFunction(31), true);
    assertEqual(primeFunction(97), true);
});

// Test larger composites
test("Larger composites: 21, 25, 49, 77, 91", () => {
    assertEqual(primeFunction(21), false); // 3×7
    assertEqual(primeFunction(25), false); // 5×5
    assertEqual(primeFunction(49), false); // 7×7
    assertEqual(primeFunction(77), false); // 7×11
    assertEqual(primeFunction(91), false); // 7×13
});

console.log("\nAll tests completed!");

// Export for potential use with other test frameworks
if (typeof module !== 'undefined' && module.exports) {
    module.exports = primeFunction;
}

File: lesson_04/devynbenson/test_prime_function.py
import unittest

def prime_function(n: int) -> bool:
    if n <= 1:
        return False

    if n <= 3:
        return True

    if n % 2 == 0 or n % 3 == 0:
        return False
    
    # Check for divisors from 5 to √n
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    
    return True


class TestPrimeFunction(unittest.TestCase):
    
    def test_edge_cases(self):
        """Test edge cases: numbers <= 1"""
        self.assertFalse(prime_function(1))  # only one positive divisor
        self.assertFalse(prime_function(0))
        self.assertFalse(prime_function(-1))
        self.assertFalse(prime_function(-5))
    
    def test_small_primes(self):
        """Test small prime numbers"""
        self.assertTrue(prime_function(2))   # only even prime
        self.assertTrue(prime_function(3))
        self.assertTrue(prime_function(5))
        self.assertTrue(prime_function(7))
        self.assertTrue(prime_function(11))
        self.assertTrue(prime_function(13))
    
    def test_composite_numbers(self):
        """Test composite numbers"""
        self.assertFalse(prime_function(4))  # 4 = 2×2, composite
        self.assertFalse(prime_function(6))  # 6 = 2×3, composite
        self.assertFalse(prime_function(8))
        self.assertFalse(prime_function(9))  # 9 = 3×3, composite
        self.assertFalse(prime_function(10))
        self.assertFalse(prime_function(12))
        self.assertFalse(prime_function(15))
    
    def test_larger_primes(self):
        """Test larger prime numbers"""
        self.assertTrue(prime_function(17))
        self.assertTrue(prime_function(19))
        self.assertTrue(prime_function(23))
        self.assertTrue(prime_function(29))
        self.assertTrue(prime_function(31))
        self.assertTrue(prime_function(97))
    
    def test_larger_composites(self):
        """Test larger composite numbers"""
        self.assertFalse(prime_function(21))  # 3×7
        self.assertFalse(prime_function(25))  # 5×5
        self.assertFalse(prime_function(49))  # 7×7
        self.assertFalse(prime_function(77))  # 7×11
        self.assertFalse(prime_function(91))  # 7×13


if __name__ == '__main__':
    # Run the tests
    unittest.main()
